pkg/controllers/node/health/controller.go

```diff
@@ -130,3 +141,38 @@ func (c *Controller) annotateTerminationGracePeriod(ctx context.Context, nodeCla
 
 	return nil
 }
+
+// Consider a nodepool unhealthy if less then 20% of the nodes owned by the nodepool have an unhealthy condition
+// defined by the cloud provider
+func (c *Controller) isNodePoolHealthy(ctx context.Context, nodeClaim *v1.NodeClaim) (bool, error) {
+	nodePoolName, found := nodeClaim.Labels[v1.NodePoolLabelKey]
+	if !found {
+		return false, fmt.Errorf("nodepool cannot be found from node")
+	}
+
+	nodeList := &corev1.NodeList{}
+	if err := c.kubeClient.List(ctx, nodeList, client.MatchingLabels(map[string]string{v1.NodePoolLabelKey: nodePoolName})); err != nil {
+		return false, err
+	}
+
+	for _, policy := range c.cloudProvider.RepairPolicies() {
+		unhealthyNodeCount := lo.CountBy(nodeList.Items, func(node corev1.Node) bool {
+			nodeCondition := nodeutils.GetCondition(lo.ToPtr(node), policy.ConditionType)
+			return nodeCondition.Status == policy.ConditionStatus
+		})
+
+		// This will round up to the nearest whole number. Therefore, a unhealthy node can
+		// sometimes exceed the 20% limit. This is the same as how Kubernetes
+		// handles MaxUnavailable with PDBs. Take the case with 20% unhealthy node threshold, but
+		// 3 nodes. Karpenter will opt to allow 1 node unhealthy nodepool, rather than
+		// blocking all node repair for this nodepool.
```

Comment:
nit: would probably be good to move this up to the method comment since this is part of the definition for "what makes a nodepool healthy".